### PR TITLE
give better suggestion for unused pattern

### PR DESCRIPTION
**Pull request by martijnhoekstra**
re-submission of https://github.com/scala/scala/pull/8377 of which the github interface seems broken - I can't re-open and it's not picking up the new commit.

### CODE REVIEW

**Review by hrhino — APPROVED** [review]
SGTM.

Github doesn't let you reopen PRs if you've force-pushed to the branch while it was closed.

File: test/files/neg/unused.check [hunk 1]
```diff
@@ -0,0 +1,21 @@
+unused.scala:7: warning: pattern var a in value patvars is never used: use a wildcard `_` or suppress this warning with `a@_`
+    case C(a, _, _, _) => 1
+           ^
+unused.scala:8: warning: pattern var b in value patvars is never used: use a wildcard `_` or suppress this warning with `b@_`
+    case C(_, b, c@_, _) => 2
+              ^
+unused.scala:9: warning: pattern var a in value patvars is never used: use a wildcard `_` or suppress this warning with `a@_`
+    case C(a, _, c, d@_) => 3
```
Comment (hrhino) [inline test/files/neg/unused.check]:
```suggestion
    case C(a, b@_, c, d@_) => 3
```

among others

**Review by hrhino — CHANGES_REQUESTED** [review]
I forgot to ask Jenkins for his SGTM.

File: test/files/neg/unused.check (hunk 1 above)
Comment (martijnhoekstra) [reply test/files/neg/unused.check]: That's the darndest thing, I created it with update--check

**Comment by martijnhoekstra** [comment]
CI failure seems spurious. Is it?

File: test/files/neg/unused.check [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+           ^
+unused.scala:9: warning: pattern var c in value patvars is never used: use a wildcard `_` or suppress this warning with `c@_`
+    case C(a, b@_, c, d@_) => 3
+                   ^
+unused.scala:10: warning: pattern var e in value patvars is never used: use a wildcard `_` or suppress this warning with `e@_`
+    case e => 4
+         ^
+unused.scala:8: warning: unreachable code
+    case C(_, b, c@_, _) => 2
+                            ^
```
Comment (hrhino) [inline test/files/neg/unused.check:18]: I wonder why this is the only unreachability warning.

**Review by hrhino — APPROVED** [review]

File: test/files/neg/unused.check (hunk 2 above)
Comment (martijnhoekstra) [reply test/files/neg/unused.check:18]: Probably at most one unreachable warning per match expression?

**Pull request merged**